### ISSUE TITLE
Module: Video UI Mediator

### ISSUE THREAD

**Issue by mikkoh** [Jam3/tech-we-want-to-use#3]
Dev Time: 3 days

A module which does not include any ui itself but is used for ui to hook into to control video playback:  pause, play, seek, etc. It should also handle eventing such as preloading and progress of a video.

**Comment by miguelmoraleda** [comment]
I did this: 

https://www.npmjs.com/package/youtube-iframe-player

**Comment by mikkoh** [comment]
Has anyone used this yet? It might be good to add in unit tests.

**Comment by mikkoh** [comment]
@miguelmoraleda this seems to be for youtube. @njam3 didnt you do something like this?

**Comment by njam3** [comment]
Yup, https://www.npmjs.com/package/video-element
works for HTML5 or YouTube with the same API.

**Comment by mikkoh** [comment]
What are the differences between what @miguelmoraleda did and yours @njam3 ?

**Comment by njam3** [comment]
@MikkoH Mine handles both HTML5 and YouTube Video with the same interface. There has been talks about added a flash version too. Think of it as a lightweight VideoJS.

**Comment by mikkoh** [comment]
I just wonder if the same "interface"/blueprint should be used but it would be two different modules. Because lets say you want to implement it for vimeo next, or like audio files or something.

**Comment by njam3** [comment]
I don't see Audio fitting into this player, but the app requires different files based on the type of video you choose, so adding vimeo, or even flash, would be pretty easy.

**Comment by mikkoh** [comment]
But the same events and functions are used to control audio too. Can you show an example of what you mean by "app requires different files based on the type of video you choose"?

**Comment by jmckinnell** [comment]
Using WebAudioAPI and/or HTML5 audio is very different from controlling a Video element from a functional standpoint. There are a lot of extra considerations to make, that go outside the scope of a Video player.

**Comment by njam3** [comment]
https://github.com/Jam3/video-element/tree/master/lib

I choose which one to require based on the type they pick.

**Comment by mikkoh** [comment]
Hmmm... I wonder if it'd be better to do:

```
{
  type: require('video-element-youtube')
}
```

Sort of like the way passport does it. http://passportjs.org/

**Comment by mattdesl** [comment]
@MikkoH 
I think we should try to avoid that style for small modules; it creates a peer dependency and leads to a lot of versioning bugs/subtleties down the road. It's also really hard to test a module against different versions of another module. 

Another way to tackle it is to develop each video player independently (`video-element-youtube` etc). Then `video-element` wraps them all up together in the same module and provides a consistent interface, without needing any peer dependencies.

**Comment by mikkoh** [comment]
@mattdesl i may have conveyed myself incorrectly. Along with my suggestion I meant to suggest creating separate modules for each player.

But yeah it might not be a bad to keep it all together and if it becomes too big then separate it. But it might be bad to do something like.

```
{
  type: require('video-element/youtube')
}
```

This way file size stays lower. Otherwise if lets say you only want to use the default `video-element` then youtube, vimeo, etc would be compiled into the source.

**Comment by mattdesl** [comment]
Yeah. I think we should just avoid passing one module to another module. e.g.

``` js
// prone to peer dependency issues
require('video-element')({
  type: require('video-element-youtube')
})
```
